modules/shredder/src/test/scala/com/snowplowanalytics/snowplow/rdbloader/shredder/batch/good/widerow/WideRowParquetCustomUnstructSpec.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) 2012-2021 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at
+ * http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and
+ * limitations there under.
+ */
+package com.snowplowanalytics.snowplow.rdbloader.shredder.batch.good.widerow
+
+import com.snowplowanalytics.snowplow.analytics.scalasdk.Event
+import com.snowplowanalytics.snowplow.rdbloader.common.config.ShredderConfig.Formats.WideRow
+import com.snowplowanalytics.snowplow.rdbloader.shredder.batch.ShredJobSpec
+import com.snowplowanalytics.snowplow.rdbloader.shredder.batch.ShredJobSpec._
+import org.apache.spark.sql.types._
+import org.specs2.mutable.Specification
+
+class WideRowParquetCustomUnstructSpec extends Specification with ShredJobSpec {
+  override def appName = "wide-row"
+  sequential
+  "A job which is configured for wide row parquet output" should {
+    val inputEvents = readResourceFile(ResourceFile("/widerow/parquet/input-events-custom-unstruct"))
+    runShredJob(events = ResourceFile("/widerow/parquet/input-events-custom-unstruct"), wideRow = Some(WideRow.Parquet))
+
+    "transform the enriched event to wide row parquet" in {
+      val lines = readParquetFile(spark, dirs.goodRows)
+        .toSet
+      val expected = inputEvents
+        .flatMap(Event.parse(_).toOption)
+        .map(testEventUpdate)
+        .map(_.toJson(true).deepDropNullValues)
+        .toSet
+      lines.size must beEqualTo(100)
+      lines must beEqualTo(expected)
```

Comment:
This is a good test!  I like it.

I can think of edge cases you are not yet testing.  For example,

In schema `1-1-0`, `f_field` is allowed to be a string or an int, but in your data set it is always a string.  You might need a different way to test the case when the data set contains integer and string values of `f_field` for version `1-1-0`.  Because if you naively assert `lines must beEqualTo(expected)` then it will (correctly) fail, because we cast the numbers to string when we write to parquet.

Your schema `1-1-0` has `additionalProperties: true` but I think you never set any additional properties.  Again, the assertion `lines must beEqualTo(expected)` would correctly fail, which makes this test more tricky.

**More importantly:**

There is another set of tests that are realistic and yet error-prone.  It is important we test this scenario somehow:

Step 1: Transform some data containing schema `1-0-0` only.  Output file name `file1.parquet`.
Step 2: Transform some data containing schema `1-0-1` only.  Output file name `file2.parquet`
Step 3: Test that `file1.parquet` and `file2.parquet` can both be loaded into the same databricks table.  Which I think is equivalent to saying..... Test that spark can read `file1.parquet` and `file2.parquet` if you [enable schema merging](https://spark.apache.org/docs/latest/sql-data-sources-parquet.html#schema-merging).

Repeat those steps for lots of combinations of schema and data.

If we can show those tests passing then I will have much more confidence in our ability to load databricks.

Reply:
You are right about first two points. I will look what I can do in there.

I've already made loading tests similar to what you described. I've prepared batches like following:
* First batch only containing data for `1-0-0`
* Second batch contains  data for `1-0-0` and `1-0-1`.
* Third batch contains data for `1-0-0`, `1-0-1` and `1-0-2`
*  and so on ...
It is a bit different than what you described since next batches contain schemas incrementally instead of containing only one version but I guess it is not problem in this case since parquet schema will be constructed according to latest one in a model. I can also make some tests like how you described. 

Visibility of those tests is other problem. I guess it would be better if I put test batches onto some location and explain how tests can be repeat and the test result in some document.